RZDataBinding/RZDBCoalesce.m

```diff
@@ -0,0 +1,309 @@
+{
+    NSUInteger existingIdx = [self.notifications indexOfObjectPassingTest:^BOOL(RZDBNotification *existing, NSUInteger idx, BOOL *stop) {
+        return [notification isEqual:existing];
+    }];
+
+    if ( existingIdx != NSNotFound ) {
+        RZDBNotification *existing = self.notifications[existingIdx];
+
+        if ( notification.changeDict[kRZDBChangeKeyNew] != nil ) {
+            existing.changeDict[kRZDBChangeKeyNew] = notification.changeDict[kRZDBChangeKeyNew];
+        }
+        else {
+            [existing.changeDict removeObjectForKey:kRZDBChangeKeyNew];
+        }
+    }
+    else {
+        [self.notifications addObject:notification];
+    }
+}
+
+@end
+
+#pragma mark - RZDBNotification implementation
+
+@implementation RZDBNotification
+
++ (instancetype)notificationWithInvocation:(NSInvocation *)invocation
+{
+    RZDBNotification *notification = nil;
+
+    NSMethodSignature *methodSig = invocation.methodSignature;
+
+    // rzdb callbacks must have void return type, and
+    // if there are any arguments, the first and only argument must be a dictionary
+    if ( methodSig.methodReturnLength == 0 && (methodSig.numberOfArguments == 2 || methodSig.numberOfArguments == 3) ) {
+        notification = [[self alloc] init];
+        notification.target = invocation.target;
+        notification.action = invocation.selector;
+
+        if ( methodSig.numberOfArguments == 3 && strcmp([methodSig getArgumentTypeAtIndex:2], @encode(id)) == 0 ) {
+            __unsafe_unretained id arg1;
+            [invocation getArgument:&arg1 atIndex:2];
+
+            if ( [arg1 isKindOfClass:[NSDictionary class]] ) {
+                notification.changeDict = [arg1 mutableCopy];
+            }
+            else {
+                notification = nil;
+            }
+        }
+    }
+
+    return notification;
+}
+
+- (BOOL)isEqual:(id)object
+{
+    BOOL equal = [super isEqual:object];
+
+    if ( !equal && [object isKindOfClass:[RZDBNotification class]] ) {
+        RZDBNotification *other = (RZDBNotification *)object;
+
+        equal = (self.target == other.target) &&
+        (self.action == other.action) &&
+        (self.changeDict == other.changeDict ||
+         (self.changeDict[kRZDBChangeKeyObject] == other.changeDict[kRZDBChangeKeyObject] &&
+          [self.changeDict[kRZDBChangeKeyKeyPath] isEqualToString:other.changeDict[kRZDBChangeKeyKeyPath]]
+          )
+         );
+    }
+
+    return equal;
+}
+
+- (void)send
+{
+    if ( self.target != nil && self.action != NULL ) {
+        if ( self.changeDict != nil ) {
+            ((void(*)(id, SEL, NSDictionary *))objc_msgSend)(self.target, self.action, [self.changeDict copy]);
+        }
+        else {
+            ((void(*)(id, SEL))objc_msgSend)(self.target, self.action);
+        }
+    }
+}
+
+@end
+
+#pragma mark - NSObject+RZDBCoalesce implementation
+
+@implementation NSObject (RZDBCoalesce)
+
+- (id)rz_coalesceProxy
+{
+    id proxy = objc_getAssociatedObject(self, _cmd);
```

Comment:
Can't have the proxy hold onto the coalesce since it's thread local. Should be able to have the proxy have a strong reference, and just return a new one every time.